src/names/scopes.md

```diff
@@ -1,3 +1,329 @@
 # Scopes
 
-> **Note**: This is a placeholder for future expansion.
+A *scope* is the region of source text where a named [entity] may be referenced with that name.
+The following sections provide details on the scoping rules and behavior, which depend on the kind of entity and where it is declared.
+The process of how names are resolved to entities is described in the [name resolution] chapter.
+More information on "drop scopes" used for the purpose of running destructors maybe be found in the [destructors] chapter.
+
+## Item scopes
+
+The name of an [item][items] declared directly in a [module] has a scope that extends from the start of the module to the end of the module. These items are also members of the module and can be referred to with a [path] leading from their module.
+
+The name of an item declared as a [statement] has a scope that extends from the start of the block the item statement is in until the end of the block.
+
+It is an error to introduce an item with a duplicate name of another item in the same [namespace] within the same module or block.
+[Asterisk glob imports] have special behavior for dealing with duplicate names and shadowing, see the linked chapter for more details.
+Items in a module may shadow items in a [prelude](#prelude-scopes).
+
+Item names from outer modules are not in scope within a nested module.
+A [path] may be used to refer to an item in another module.
+
+### Associated item scopes
+
+[Associated items] are not scoped and can only be referred to by using a [path] leading from the type they are associated with.
+[Methods] can also be referred to via [call expressions].
+
+Similar to items within a module or block, it is an error to introduce an item within a trait or implementation that is a duplicate of another item in the trait or impl in the same namespace.
+
+## Pattern binding scopes
+
+The scope of a local variable [pattern] binding depends on where it is used:
+
+* [`let` statement] bindings range from just after the `let` statement until the end of the block where it is declared.
+* [Function parameter] bindings are within the body of the function.
+* [Closure parameter] bindings are within the closure body.
+* [`for`] and [`while let`] bindings are within the loop body.
+* [`if let`] bindings are within the consequent block.
```

Comment:
is "consequent" a technical term here? Trying to figure out whether this implies that it goes out of scope in the `else` block.

Reply:
Yea, the consequent block is the block immediately after the conditional expression. It is defined in https://doc.rust-lang.org/nightly/reference/expressions/if-expr.html.  Else blocks are called "else blocks".